catalog/rocks/catalog.go

```diff
@@ -0,0 +1,244 @@
+	DiffTypeRemoved
+	DiffTypeChanged
+	DiffTypeConflict
+)
+
+type Diff struct {
+	Path Path
+	Type DiffType
+}
+
+// function/methods receiving the following basic types could assume they passed validation
+type (
+	// StorageNamespace is the URI to the storage location
+	StorageNamespace string
+
+	// RepositoryID is an identifier for a repo
+	RepositoryID string
+
+	// Path represents a logical path for an entry
+	Path string
+
+	// Ref could be a commit ID, a branch name, a Tag
+	Ref string
+
+	// TagID represents a named tag pointing at a commit
+	TagID string
+
+	// CommitID is a content addressable hash representing a Commit object
+	CommitID string
+
+	// BranchID is an identifier for a branch
+	BranchID string
+
+	// TreeID represents a snapshot of the tree, referenced by a commit
+	TreeID string
+
+	// StagingToken represents a namespace for writes to apply as uncommitted
+	StagingToken string
+
+	// CommonPrefix represents a path prefixing one or more Entry objects
+	CommonPrefix string
+)
+
+// Listing represents either an entry or a CommonPrefix
+type Listing struct {
+	CommonPrefix
+	*Entry
+}
+
+// Interfaces
+type Catalog interface {
+	// entries
+	GetEntry(ctx context.Context, repositoryID RepositoryID, ref Ref, path Path) (*Entry, error)
+	SetEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path, entry Entry) error
+	DeleteEntry(ctx context.Context, repositoryID RepositoryID, branchID BranchID, path Path) error
+	ListEntries(ctx context.Context, repositoryID RepositoryID, ref Ref, prefix, from, delimiter string, amount int) ([]Listing, bool, error)
+
+	// refs
+	CreateBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) (Branch, error)
+	GetBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) (Branch, error)
+	Dereference(ctx context.Context, repositoryID RepositoryID, ref Ref) (CommitID, error)
+	Log(ctx context.Context, repositoryID RepositoryID, commitID CommitID, amount int) ([]Commit, bool, error)
+	ListBranches(ctx context.Context, repositoryID RepositoryID, from BranchID, amount int) ([]Branch, bool, error)
+	DeleteBranch(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+
+	// commits
+	Commit(ctx context.Context, repositoryID RepositoryID, branchID BranchID, commit Commit) (CommitID, error)
+	Reset(ctx context.Context, repositoryID RepositoryID, branchID BranchID) error
+	Revert(ctx context.Context, repositoryID RepositoryID, branchID BranchID, ref Ref) error
+
+	// diffs and merges
+	Merge(ctx context.Context, repositoryID RepositoryID, from Ref, to BranchID) (CommitID, error)
+	DiffUncommitted(ctx context.Context, repositoryID RepositoryID, branchID BranchID, from Path, amount int) ([]Diff, bool, error)
+	Diff(ctx context.Context, repositoryID RepositoryID, left, right Ref, from Path, amount int) ([]Diff, bool, error)
+}
+
+// internal structures used by Catalog
+type RepositoryIterator interface {
```

Comment:
Let's document the expected behaviour for each one of the First, Next, SeekGE , pending the iterator discussion.